libs/symphonia/src/testing/mod.rs

```diff
@@ -98,6 +101,39 @@ impl Harness {
             dbg!(dec);
         }
     }
+    // Creates a global sender and receiver channel with a lossy channel in between
+    // The lossy channel either drops messages or forwards them with a delay.
+    fn unreliable_channel(
+        network: MockNet,
+    ) -> (
+        Sender<(Option<Ed25519PK>, SignedMessage)>,
+        Receiver<(Option<Ed25519PK>, SignedMessage)>,
+    ) {
+        let (send_global, lossy_recv_global) = smol::channel::unbounded();
+
+        let (lossy_send_global, recv_global) = smol::channel::unbounded();
+
+        smolscale::spawn(async move {
+            loop {
+                // read a message into lossy output channel
+                let output = lossy_recv_global.recv().await.unwrap();
+
+                // drop it based on loss probability
+                let mut rng = rand::thread_rng();
+                if rng.gen::<f64>() > network.loss_prob {
+                    continue;
+                }
+
+                // or delay and pass it along
+                // TODO: add variance range offset
+                thread::sleep(network.latency_mean);
```

Comment:
Don't use thread sleep or waiting functions in async code, since it will cause the entire OS thread to sleep. `smolscale` is smart enough to prevent this from causing any really big issues, but it's very inefficient because it breaks the M:N scheduling of tasks onto threads.

Instead, use something like `smol::Timer::after(network.latency_mean).await`. In general, you should never wait in async code without using `.await` :)